src/acl.erl

```diff
@@ -26,101 +26,96 @@
--spec match_rule(mongooseim:host_type_or_global(), jid:lserver(), rule(), jid:jid(),
+-spec match_rule(mongooseim:host_type_or_global(), jid:lserver(), rule_name(), jid:jid(),
                  Default :: acl_result()) ->
           acl_result().
-match_rule(global, Domain, Rule, JID, Default) ->
-    case mongoose_config:lookup_opt({access, Rule, global}) of
+match_rule(HostType, Domain, RuleName, JID, Default) ->
+    case mongoose_config:lookup_opt([{access, HostType}, RuleName]) of
         {error, not_found} ->
             Default;
-        {ok, ACLs} ->
-            match_acls(ACLs, JID, global, Domain)
-    end;
-match_rule(HostType, Domain, Rule, JID, Default) ->
-    case {mongoose_config:lookup_opt({access, Rule, global}),
-          mongoose_config:lookup_opt({access, Rule, HostType})} of
-        {{error, not_found}, {error, not_found}} ->
-            Default;
-        {{error, not_found}, {ok, HostACLs}} ->
-            match_acls(HostACLs, JID, HostType, Domain);
-        {{ok, GlobalACLs}, {error, not_found}} ->
-            match_acls(GlobalACLs, JID, HostType, Domain);
-        {{ok, GlobalACLs}, {ok, HostACLs}} ->
-            match_acls(merge_acls(GlobalACLs, HostACLs), JID, HostType, Domain)
+        {ok, RuleSpec} ->
+            match_rule_clauses(RuleSpec, JID, HostType, Domain)
     end.
 
--spec merge_acls([{any(), rule()}], [{any(), rule()}]) -> [{any(), rule()}].
-merge_acls(Global, HostLocal) ->
+%% Merge host-type rules with global ones on startup
+-spec merge_access_rules([rule_clause()], [rule_clause()]) -> [rule_clause()].
+merge_access_rules(Global, HostLocal) ->
     case lists:reverse(Global) of
-        [{allow, all} | Rest] ->
-            lists:reverse(Rest) ++ HostLocal ++ [{allow, all}];
+        [#{acl := all, value := allow} = AllowAll | Rest] ->
+            lists:reverse(Rest) ++ HostLocal ++ [AllowAll];
         _ ->
             Global ++ HostLocal
     end.
 
--spec match_acls(ACLs :: [{any(), rule()}], jid:jid(), mongooseim:host_type_or_global(),
-                 jid:lserver()) ->
+-spec match_rule_clauses([rule_clause()], jid:jid(), mongooseim:host_type_or_global(),
+                         jid:lserver()) ->
           acl_result().
-match_acls([], _, _HostType, _Domain) ->
+match_rule_clauses([], _, _HostType, _Domain) ->
     deny;
-match_acls([{Value, Rule} | ACLs], JID, HostType, Domain) ->
-    case match_acl(Rule, JID, HostType, Domain) of
+match_rule_clauses([#{acl := ACLName, value := Value} | RemainingClauses], JID, HostType, Domain) ->
+    case match_acl(ACLName, JID, HostType, Domain) of
         true ->
             Value;
         _ ->
-            match_acls(ACLs, JID, HostType, Domain)
+            match_rule_clauses(RemainingClauses, JID, HostType, Domain)
     end.
 
--spec match_acl(rule(), jid:jid(), mongooseim:host_type_or_global(), jid:lserver()) -> boolean().
+-spec match_acl(acl_name(), jid:jid(), mongooseim:host_type_or_global(), jid:lserver()) ->
+          boolean().
```

Comment:
```suggestion
                boolean().
```

Reply:
It's correct as it is. After your suggestion it would look as if `boolean` was another argument. Anyway, the Erlang mode for Emacs auto-indents it this way.